### DISCUSSION TITLE
Module not found errors

### DISCUSSION THREAD

**Discussion by efiw** [panva/jose#151] [Q&A]
Hey all, 

I've added jose (3.5.1) as a dependency to my typescript project, getting the following error when compiling:

```
Error: Cannot find module 'jose/jwks/remote'
```

when I try to use `import createRemoteJWKSet from 'jose/jwks/remote';`

Has anyone encountered it before?

Thanks

**Comment by panva** [comment]
Upgrade typescript, make sure your node.js is up to date. Make sure your babels, webpacks etc support subpath exports.

**Reply by panva** [reply]
And if all else fails, go for the dist files directly.

**Reply by panva** [reply]
Also try [this](https://github.com/panva/jose/issues/143#issuecomment-759423039), but report back your results and way more details about your setup. I wanna know what makes it not work for you and track its progress in supporting this feature of package.json in node.

**Reply by efiw** [reply]
solved by upgrading node version, thanks!

**Comment by jonathanstanley** [comment]
I had this same error with `jose/jwe/compact/encrypt`. I'm building with gatsby (with their [webpack](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/utils/webpack.config.js)) and node 12.19.0. In my case I may be stuck with node12 for a bit, but tried updating node to 15.2.0 just to see... didn't help.  Per https://github.com/panva/jose/issues/143, I was hoping the experimental library would avoid the `can't resolve...` issue

in my case, the experimental produced : 
 ```
ERROR #98123  WEBPACK
undefined failed
Unexpected token (26:9)
File: node_modules/jose/dist/node/esm/index.js:26:9
````

what is strange to me is subpaths work with other packages.  ex:
```javascript
import Avatar from "@material-ui/core/Avatar";
import Typography from "@material-ui/core/Typography";
```

@panva thank you very much for your library!  I hope this info helps.

**Reply by panva** [reply]
@jonathanstanley webpack 4 does not support subpath exports, webpack 5 does. gatsby should [upgrade their webpack](https://github.com/gatsbyjs/gatsby/pull/29145) bundle to accomodate new features.

`@material-ui/core/Avatar` is not a subpath export, its a directly nested file.

**Reply by panva** [reply]
ad experimental, it actually helped - since it located an actual dist folder file, but the unexpected token signals a different issue. I'm not going to investigate further. All these tools that include proprietary loaders should just get in line and use the one included in the runtime.

**Reply by panva** [reply]
As a workaround, locate the file to load with the right package flavour yourself. e.g. `jose/dist/node/esm/jwt/sign.js`. That's essentially what subpath exports do for you, that makes this a truly universal library since the same code would work in a browser bundle as well as node despite each using a different crypto runtime.

**Reply by jonathanstanley** [reply]
thanks!  i actually just tried that `import CompactEncrypt from "jose/dist/browser/jwe/compact/encrypt";
`

but got

```
1 | import crypto from './webcrypto.js';
> 2 | const random = crypto.getRandomValues.bind(crypto);
    |                      ^
  3 | export default random;
  4 | 


  WebpackError: TypeError: Cannot read property 'getRandomValues' of undefined
```

**Reply by panva** [reply]
But your target runtime isn't the browser now, is it?

**Reply by jonathanstanley** [reply]
ah, i noticed you suggested the node/esm. i did point to that and it did compile.  though i'm not quite sure if this will work since it is ultimately compiling a web app.

**Reply by panva** [reply]
(╯°□°)╯︵ ┻WEBPACKv4┻

**Reply by panva** [reply]
https://github.com/gatsbyjs/gatsby/pull/29145

**Reply by jonathanstanley** [reply]
lol understood!

**Reply by panva** [reply]
works flawlessly - https://github.com/panva/jose/issues/111

**Reply by jonathanstanley** [reply]
just wanted to confirm: **webpack v4 was the issue**.

gatsby updated to webpack v5 and i got jose working very quickly.  thanks @panva !

**Comment by panva** [comment]
I'm done waiting for these ecosystems to catch up. jose v4.x will be coming out today/tomorrow with only top level named exports that solve all these tools' problems.
